database/rrdhost.c

```diff
@@ -737,25 +706,17 @@ int rrd_init(char *hostname, struct rrdhost_system_info *system_info) {
         return 1;
     }
 
-#ifdef ENABLE_DBENGINE
-    char dbenginepath[FILENAME_MAX + 1];
-    int ret;
-    snprintfz(dbenginepath, FILENAME_MAX, "%s/dbengine", localhost->cache_dir);
-    ret = mkdir(dbenginepath, 0775);
-    if (ret != 0 && errno != EEXIST)
-        error("Host '%s': cannot create directory '%s'", localhost->hostname, dbenginepath);
-    else  // Unconditionally create multihost db to support on demand host creation
-        ret = rrdeng_init(NULL, NULL, dbenginepath, default_rrdeng_page_cache_mb, default_multidb_disk_quota_mb);
-    if (ret) {
-        error(
-            "Host '%s' with machine guid '%s' failed to initialize multi-host DB engine instance at '%s'.",
-            localhost->hostname, localhost->machine_guid, localhost->cache_dir);
-        rrdhost_free(localhost);
-        localhost = NULL;
-        rrd_unlock();
-        fatal("Failed to initialize dbengine");
+    for (STORAGE_ENGINE* eng = storage_engine_foreach_init(); eng; eng = storage_engine_foreach_next(eng)) {
+        if (eng->api.engine_ops.create) {
+            STORAGE_ENGINE_INSTANCE* ret = storage_engine_new(eng, localhost);
```

Comment:
What is the intention here from the API's perspective? Reading this, I'd say that we are creating a new instance for every engine that supports the `create` OP. IOW, if X engines support the `create` operation, we'll create X storage engine instances.

Reply:
Indeed. This allow initializing the global context of supported storage engines.
This is required because child nodes can be configured to use different engines, and these engines might use a global context.
This behavior is derived from the current behavior (multidb_ctx is initialized even if dbengine is not configured for this host)

Reply:
This was wrapped under an `ifdef ENABLE_DBENGINE`, now it will run for every single memory mode. My question is what is (or will be) the intention when we have multiple different storage_engines that support the `create` op. IOW, what will happen at a functional level when this will run for both the dbengine & the mongodb storage engines?

Reply:
`ifdef ENABLE_DBENGINE` is now in `storage_engine.c`, with the same result : dbengine is not initialized if netdata is built without dbengine support. In this case the iteration would not get though dbengine because dbengine would not be in the storage engine list.

Multiple engines can each use a global context. For instance, dbengine and mongoengine could each have their own context. Some child nodes could use dbengine and other children mongoengine (that's allowed by netdata configuration).

Functionally, contexts are initialized/destroyed at two levels (like the current design):
* globally during initialization and shutdown
* for each host, when created and destroyed

This gives the opportunity to storage engines to chose when they need a new context, or to use different opaque structures for the global context and host contexts (which allows mongoengine to have per-host caches for instance).

Reply:
Hi @aberaud! I have run few tests and it starts to look nice.

I agree with @vkalintiris that this point needs a bit more attention to ensure that is scalable and simple in cases with multiple memory modes(>2) that support the create function. 

I was trying to think what will happen,
1. If the create function is responsible to assign the storage engine instance to the host storage engine context then does it mean that the localhost will always start with the storage engine at the end of the global list (do we want this)?
2. Where are the redundant (not assigned to localhost/child hosts but created) memory global storage engine instances are being freed/destroyed?

Reply:
1. Just updated the PR to account for this scenario. We now only attach the context to the host if this host is configured for this engine (based on `host->rrd_memory_mode`).
2. Current logic is preserved : when a host is destroyed, `engine_ops.exit` and `engine_ops.destroy` are only called if the context object is different from the global object. During global shutdown, in `netdata_cleanup_and_exit`, `exit` and `destroy` are call on remaining engine global contexts.